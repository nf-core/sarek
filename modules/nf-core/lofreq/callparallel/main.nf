process LOFREQ_CALLPARALLEL {
    tag "$meta.id"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/lofreq:2.1.5--py38h588ecb2_4' :
        'biocontainers/lofreq:2.1.5--py38h588ecb2_4' }"

    input:
    tuple val(meta) , path(bam), path(bai), path(intervals)
    tuple val(meta2), path(fasta)
    tuple val(meta3), path(fai)

    output:
    tuple val(meta), path("*.vcf.gz")    , emit: vcf
    tuple val(meta), path("*.vcf.gz.tbi"), emit: tbi
    path "versions.yml"                  , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def options_intervals = intervals ? "-l ${intervals}" : ""

    def alignment_cram =  bam.Extension == "cram" ? true : false
    def alignment_bam = bam.Extension == "bam" ? true : false
    def alignment_out = alignment_cram ? bam.BaseName + ".bam" : "${bam}"

    def samtools_cram_convert = ''
    samtools_cram_convert += alignment_cram ? "    samtools view -T ${fasta} ${bam} -@ $task.cpus -o ${alignment_out}\n" : ''
    samtools_cram_convert += alignment_cram ? "    samtools index ${alignment_out}\n" : ''

    def samtools_cram_remove = ''
    samtools_cram_remove += alignment_cram ? "    rm ${alignment_out}\n" : ''
    samtools_cram_remove += alignment_cram ? "    rm ${alignment_out}.bai\n " : ''
    """
    $samtools_cram_convert

    lofreq \\
        call-parallel \\
        --pp-threads $task.cpus \\
        $args \\
        $options_intervals \\
        -f $fasta \\
        -o ${prefix}.vcf.gz \\
        $alignment_out

    $samtools_cram_remove

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        lofreq: \$(echo \$(lofreq version 2>&1) | sed 's/^version: //; s/ *commit.*\$//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    echo "" | gzip > ${prefix}.vcf.gz
    echo "" | gzip > ${prefix}.vcf.gz.tbi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        lofreq: \$(echo \$(lofreq version 2>&1) | sed 's/^version: //; s/ *commit.*\$//')
    END_VERSIONS
    """
}
