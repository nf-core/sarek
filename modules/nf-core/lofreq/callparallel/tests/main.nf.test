nextflow_process {

    name "Test Process LOFREQ_CALLPARALLEL"
    script "../main.nf"
    process "LOFREQ_CALLPARALLEL"

    tag "modules"
    tag "modules_nfcore"
    tag "lofreq"
    tag "lofreq/callparallel"

    test("sarscov2 - bam") {

        when {
            process {
                """
                input[0] = [ [ id:'test' ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true),
                        []
                        ]
                input[1] = [ [ id:'fasta' ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                        ]
                input[2] = [ [ id:'fai' ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                        ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.versions,
                    process.out.tbi,
                    path(process.out.vcf[0][1]).vcf.summary
                    ).match() },
            )
        }

    }

    test("sarscov2 - bam - bed") {

        when {
            process {
                """
                input[0] = [ [ id:'test' ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/bed/test.bed', checkIfExists: true)
                        ]
                input[1] = [ [ id:'fasta' ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                        ]
                input[2] = [ [ id:'fai' ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                        ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.versions,
                    process.out.tbi,
                    path(process.out.vcf[0][1]).vcf.summary
                    ).match() },
            )
        }

    }

    test("sarscov2 - bam - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [ [ id:'test' ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true),
                        []
                        ]
                input[1] = [ [ id:'fasta' ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                        ]
                input[2] = [ [ id:'fai' ], // meta map
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                        ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
