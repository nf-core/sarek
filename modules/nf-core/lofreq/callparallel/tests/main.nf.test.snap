{
    "sarscov2 - bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,1a60c330fb42841e8dcf3cd507a70bfc"
                    ]
                ],
                "2": [
                    "versions.yml:md5,56d45e0015add277b2689f071a4fe3e4"
                ],
                "tbi": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,1a60c330fb42841e8dcf3cd507a70bfc"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,56d45e0015add277b2689f071a4fe3e4"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.0"
        },
        "timestamp": "2024-08-28T12:01:24.268196316"
    },
    "sarscov2 - bam": {
        "content": [
            [
                "versions.yml:md5,56d45e0015add277b2689f071a4fe3e4"
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.vcf.gz.tbi:md5,4cb176febbc8c26d717a6c6e67b9c905"
                ]
            ],
            "VcfFile [chromosomes=[], sampleCount=0, variantCount=0, phased=true, phasedAutodetect=true]"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.0"
        },
        "timestamp": "2024-08-28T14:14:55.381365088"
    },
    "sarscov2 - bam - bed": {
        "content": [
            [
                "versions.yml:md5,56d45e0015add277b2689f071a4fe3e4"
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.vcf.gz.tbi:md5,4cb176febbc8c26d717a6c6e67b9c905"
                ]
            ],
            "VcfFile [chromosomes=[], sampleCount=0, variantCount=0, phased=true, phasedAutodetect=true]"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.0"
        },
        "timestamp": "2024-08-28T14:15:18.221515296"
    }
}